File: CS/UpperFunctionTest.cs
using System;
using System.Collections.Generic;
using System.Text;
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.Data.Filtering;

namespace UpperFunction {
    public class UpperFunction {
        public void Run() {
            IDataLayer dl = XpoDefault.GetDataLayer(AutoCreateOption.DatabaseAndSchema);
            UnitOfWork uof = new UnitOfWork(dl);
            uof.ClearDatabase();
            uof.UpdateSchema();
            MyObject obj1 = new MyObject(uof);
            obj1.MyCaseSensitiveProperty = "ab";
            MyObject obj2 = new MyObject(uof);
            obj2.MyCaseSensitiveProperty = "AB";
            MyObject obj3 = new MyObject(uof);
            obj3.MyCaseSensitiveProperty = "aB";
            MyObject obj4 = new MyObject(uof);
            obj4.MyCaseSensitiveProperty = "BC";
            MyObject obj5 = new MyObject(uof);
            obj5.MyCaseSensitiveProperty = "Cb";
            uof.CommitChanges();

            // Test UPPER
            CriteriaOperator filter = CriteriaOperator.Parse("Upper([MyCaseSensitiveProperty]) = ?", "AB");
            XPCollection<MyObject> collection = new XPCollection<MyObject>(uof, filter);

            // Test LOWER
            filter = CriteriaOperator.Parse("Lower([MyCaseSensitiveProperty]) = ?", "ab");
            collection = new XPCollection<MyObject>(uof, filter);

            // Test SUBSTRING
            filter = CriteriaOperator.Parse("Substring([MyCaseSensitiveProperty], 0, 1) = ?", "c");
            collection = new XPCollection<MyObject>(uof, filter);
        }
    }

    public class UpperFunctionTest : UpperFunction {
        public new void Run() {
            base.Run();
        }
    }

    class MyObject : XPObject {
        public MyObject(Session session) : base(session) { }

        string myCaseSensitiveValue;
        public string MyCaseSensitiveProperty {
            get { return myCaseSensitiveValue; }
            set { SetPropertyValue("MyCaseSensitiveProperty", ref myCaseSensitiveValue, value); }
        }
    }
}

File: VB/UpperFunctionTest.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Text
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports DevExpress.Data.Filtering

Namespace UpperFunction
	Public Class UpperFunction
		Public Sub Run()
			Dim dl As IDataLayer = XpoDefault.GetDataLayer(AutoCreateOption.DatabaseAndSchema)
			Dim uof As New UnitOfWork(dl)
			uof.ClearDatabase()
			uof.UpdateSchema()
			Dim obj1 As New MyObject(uof)
			obj1.MyCaseSensitiveProperty = "ab"
			Dim obj2 As New MyObject(uof)
			obj2.MyCaseSensitiveProperty = "AB"
			Dim obj3 As New MyObject(uof)
			obj3.MyCaseSensitiveProperty = "aB"
			Dim obj4 As New MyObject(uof)
			obj4.MyCaseSensitiveProperty = "BC"
			Dim obj5 As New MyObject(uof)
			obj5.MyCaseSensitiveProperty = "Cb"
			uof.CommitChanges()

			' Test UPPER
			Dim filter As CriteriaOperator = CriteriaOperator.Parse("Upper([MyCaseSensitiveProperty]) = ?", "AB")
			Dim collection As XPCollection(Of MyObject) = New XPCollection(Of MyObject)(uof, filter)

			' Test LOWER
			filter = CriteriaOperator.Parse("Lower([MyCaseSensitiveProperty]) = ?", "ab")
			collection = New XPCollection(Of MyObject)(uof, filter)

			' Test SUBSTRING
			filter = CriteriaOperator.Parse("Substring([MyCaseSensitiveProperty], 0, 1) = ?", "c")
			collection = New XPCollection(Of MyObject)(uof, filter)
		End Sub
	End Class

	Public Class UpperFunctionTest
		Inherits UpperFunction
		Public Shadows Sub Run()
			MyBase.Run()
		End Sub
	End Class

	Friend Class MyObject
		Inherits XPObject
		Public Sub New(ByVal session As Session)
			MyBase.New(session)
		End Sub

		Private myCaseSensitiveValue As String
		Public Property MyCaseSensitiveProperty() As String
			Get
				Return myCaseSensitiveValue
			End Get
			Set(ByVal value As String)
				SetPropertyValue("MyCaseSensitiveProperty", myCaseSensitiveValue, value)
			End Set
		End Property
	End Class
End Namespace
